pixeltable/share/packager.py

```diff
@@ -69,138 +70,122 @@ def __init__(self, table: catalog.Table, additional_md: Optional[dict[str, Any]]
-
-        # Run the select() on `self.table`, not `t`, so that we export only those rows that are actually present in
-        # `self.table`.
-        df = self.table.select(**select_exprs)
-        namespace = self.__iceberg_namespace(t)
-        self.iceberg_catalog.create_namespace_if_not_exists(namespace)
-        iceberg_schema = self.__to_iceberg_schema(df._schema)
-        iceberg_tbl = self.iceberg_catalog.create_table(f'{namespace}.{t._name}', schema=iceberg_schema)
-
-        # Populate the Iceberg table with data.
-        # The data is first loaded from the DataFrame into a sequence of pyarrow tables, batched in order to avoid
-        # excessive memory usage. The pyarrow tables are then amalgamated into the (single) Iceberg table on disk.
-        for pa_table in self.__to_pa_tables(df, actual_col_types, iceberg_schema):
-            iceberg_tbl.append(pa_table)
+        sql_types = {col.name: col.type for col in tv.store_tbl.sa_tbl.columns}
+        media_cols: set[str] = set()
+        for col in tv.cols_by_name.values():
+            if col.is_stored and col.col_type.is_media_type():
+                media_cols.add(col.store_name())
 
-    @classmethod
-    def __iceberg_namespace(cls, table: catalog.Table) -> str:
-        """
-        Iceberg tables must have a namespace, which cannot be the empty string, so we prepend `pxt` to the table path.
-        """
-        parent_path = table._parent()._path()
-        if len(parent_path) == 0:
-            return 'pxt'
-        else:
-            return f'pxt.{parent_path}'
+        parquet_schema = self.__to_parquet_schema(tv.store_tbl.sa_tbl)
+        # The parquet file naming scheme anticipates future support for partitioning.
+        parquet_dir = self.tables_dir / f'tbl_{tv.id.hex}'
+        parquet_dir.mkdir()
+        parquet_file = parquet_dir / f'tbl_{tv.id.hex}.00000.parquet'
+        _logger.info(f'Creating parquet table: {parquet_file}')
 
-    # The following methods are responsible for schema and data conversion from Pixeltable to Iceberg. Some of this
+        # Populate the Parquet table with data.
+        # The data is first loaded from the DataFrame into a sequence of pyarrow tables, batched in order to avoid
+        # excessive memory usage. The pyarrow tables are then amalgamated into the (single) Parquet table on disk.
+        # We use snappy compression for the Parquet tables; the entire bundle will be bzip2-compressed later, so
+        # faster compression should provide good performance while still reducing temporary storage utilization.
+        parquet_writer = pq.ParquetWriter(parquet_file, parquet_schema, compression='SNAPPY')
+        filter_tv = self.table._tbl_version.get()
+        row_iter = tv.store_tbl.dump_rows(tv.version, filter_tv.store_tbl, filter_tv.version)
+        for pa_table in self.__to_pa_tables(row_iter, sql_types, media_cols, parquet_schema):
+            parquet_writer.write_table(pa_table)
```

Comment:
How does this interact with Parquet compression? Does this buffer everything in memory before writing out a single parquet file? (It would need to do that for things like dictionary encoding, which parquet supports.)

Reply:
If I understand the docs right, it's partitioned into row groups of size `min(pyarrow_table_size, 2**20)`. I'm using 1000-row pyarrow tables; we can of course tweak this at any point